Assets/Mirror/Runtime/NetworkWriter.cs

```diff
@@ -19,28 +20,75 @@ public class NetworkWriter
 
         // 'int' is the best type for .Position. 'short' is too small if we send >32kb which would result in negative .Position
         // -> converting long to int is fine until 2GB of data (MAX_INT), so we don't have to worry about overflows here
-        public int Position;
-
+        int position;
         int length;
 
-        public int Length
+        public int Length => length;
+
+        public int Position
         {
-            get => length;
-            private set
+            get => position;
+            set
             {
-                EnsureCapacity(value);
-                length = value;
-                if (Position > length)
-                    Position = length;
+                position = value;
+                EnsureLength(value);
             }
         }
 
-        void EnsureCapacity(int value)
+        /// <summary>
+        /// Reset both the position and length of the stream
+        /// </summary>
+        /// <remarks>
+        /// Leaves the capacity the same so that we can reuse this writer without extra allocations
+        /// </remarks>
+        public void Reset()
         {
-            if (buffer.Length < value)
+            position = 0;
+            length = 0;
+        }
+
+        /// <summary>
+        /// Sets length, moves position if it is greater than new length
+        /// </summary>
+        /// <param name="newLength"></param>
+        /// <remarks>
+        /// Zeros out any extra length created by setlength
+        /// </remarks>
+        public void SetLength(int newLength)
+        {
+            int oldLength = length;
+
+            // use EnsureCapacity 
+            EnsureLength(newLength);
+
+            // zero out new length
+            if (oldLength < newLength)
             {
-                int capacity = Math.Max(value, buffer.Length * 2);
-                Array.Resize(ref buffer, capacity);
+                for (int i = oldLength; i < newLength; i++)
+                {
+                    buffer[i] = 0;
+                }
+                Array.Clear(buffer, oldLength, newLength - oldLength);
+            }
+
+
+            length = newLength;
+            position = Mathf.Min(position, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        void EnsureLength(int value)
```

Comment:
why 'EnsureLength'?
this function is used to increase capacity. e.g. when we have 100 bytes array but want it to grow to 200 bytes

Reply:
@vis2k previously it increased capacity.

Now it increases the length,  which may end up also increasing capacity

Reply:
why would we increase the length?
it would just fill the message with random bytes at the end.

or.. why do we need length at all? if we have position and buffer.Length then that's all we need really

Reply:
EnsureLength is used inside of NetworkWriter when writing anything to the buffer.

For example in WriteByte ` EnsureLength(position + 1);`

Most of the time it will just increase the length of the data, But if someone moves the position to an earlier then writes it just checks if the positon + new data is less than length.

Reply:
vis,  if you are at the end of the writer and write a byte you need to increase length + 1.  
Basically we always need to increase the length if we are at the end of the writer.   Sometimes we might need to increase capacity too.

Users do not manually set the length,  there is no way to end up with random bytes at all.
The only way to increase length is by writing something at the end,  so we would never end up with random bytes.